Completing the primer section on unsafe

cartermp opened the pull request:
Addressing #27.
- Added a code sample taken from CoreFx
- Linked to the folder where the source file sits
- Reworded blurb to emphasize that it should be discouraged unless
  necessary, not simply discouraged

/cc @blackdwarf @richlander @terrajobst

cartermp committed:
Completing the primer section on unsafe

* Added a code sample taken from CoreFx
* Linked to the folder where the source file sits
* Reworded blurb to emphasize that it should be discouraged unless
necessary, not simply discouraged

cartermp committed:
Adding in-document link to Unsafe Code section

blackdwarf commented:
LGTM. I was meaning to finish up the unsafe section together with the interoperability stuff, but I guess we can keep them separate as well.

richlander commented:
LGTM. The example is unsafe solely due to the APIs it calls and the one field on the struct. An additional and more expansive use of 'fixed' might be good. See the examples here: https://msdn.microsoft.com/en-us/library/f58wzh21.aspx. I suspect that there are some great examples in the BCL, maybe in something like StringBuilder.

cartermp committed:
Changing example from Linux Interop to StringBuilder

The rationale here is that people will want to see the use of unsafe
code from an algorithmic standpoint more than a systems interop
standpoint.

cartermp commented:
Updated this to use `StringBuilder`'s `ToString()` method, which makes use of `fixed` in addition to an `unsafe` block.

I would like to also highlight a code sample using some systems interop, but I also would like to keep this to just one code sample.  Is there any source you know of that uses memory manipulation in an algorithm **and** does some interop?

richlander commented on docs/concepts/primer.rst:
```diff
@@ -258,14 +259,52 @@ Unsafe Code
+unsafe code as much as possible, and test that code extremely thoroughly.
 
-TODO: Examples.
+Taken from the `StringBuilder class in the .NET BCL reference source <http://referencesource.microsoft.com/#mscorlib/system/text/stringbuilder.cs,adf60ee46ebd299f>`_:
```
This link would probably be more relevant: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Text/StringBuilder.cs

cartermp replied:
Done.

richlander commented:
LGTM modulo feedback

cartermp committed:
Change link to github source, not reference source

cartermp commented:
Changed the link.  Should be good to merge now.

richlander commented on docs/concepts/primer.rst:
```diff
@@ -258,14 +259,52 @@ Unsafe Code
+unsafe code as much as possible, and test that code extremely thoroughly.
 
-TODO: Examples.
+Taken from the `StringBuilder class in the .NET BCL source <https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Text/StringBuilder.cs#L327>`_:
```
It feels like this is missing a bit of an intro, really to clarify that this is an example of unsafe code use.

richlander commented on docs/concepts/primer.rst:
```diff
@@ -258,14 +259,52 @@ Unsafe Code
-be allowed to run in all environments. The best practice is to confine
-unsafe code as much as possible and that the vast majority of code is
-type-safe.
+arithmetic via ``unsafe`` code. These operations are needed for certain algorithms and system interoperability.  Although powerful, use of unsafe code is discouraged unless it is necessary to interop with system APIs or implement the most efficient algorithm.  Unsafe code is not guaranteed to run in all environments, and also loses the benefits of a garbage collector and type safety.  It's recomm ...
```
Do you have an example of where unsafe code doesn't run? It makes it sounds like you could build an app this way and it might unpredictably not run. It's the "not guaranteed to run" part that is a bit confusing. Do you mean "no supported"?

cartermp replied:
Yes, that's what I meant.  Fixing that!

cartermp committed:
Addressing richlander comments

cartermp committed:
Merge remote-tracking branch 'refs/remotes/dotnet/master' into unsafe-primer-sample

Conflicts:
	docs/concepts/primer.rst

cartermp committed:
Merge in updates

richlander commented on docs/concepts/primer.rst:
```diff
@@ -234,29 +235,79 @@ TODO: finish section
+^^^^^^^^^^^
+
+The CLR enables the ability to access native memory and do pointer
+arithmetic via ``unsafe`` code. These operations are needed for certain algorithms and system interoperability.  Although powerful, use of unsafe code is discouraged unless it is necessary to interop with system APIs or implement the most efficient algorithm.  Unsafe code may not predictably run in all environments, and also loses the benefits of a garbage collector and type safety.  It's recomme ...
```
I'm still wondering what you mean by "may not predictably run in all environments".

cartermp committed:
Merge remote-tracking branch 'refs/remotes/dotnet/master' into unsafe-primer-sample

cartermp committed:
Clear up what "reliably" meant in terms of execution of unsafe code

richlander commented:
Can you squash and merge? Thanks.

cartermp commented:
I dun goofed and messed things up in my environment, preventing me from merging this in properly.  Going to open a new PR with the squashed commits.
